crl/validator.go

```diff
@@ -20,346 +20,420 @@ package crl
+	v.crlChan <- func() {
+		defer close(result)
+		v.crls[endpoint] = crl
+		result <- struct{}{}
 	}
-	return result
+	<-result // make blocking call
 }
 
-func (v *validator) appendCertificates(certificates []*x509.Certificate) {
-	v.certificatesLock.Lock()
-	defer v.certificatesLock.Unlock()
+// sync tries to update all crls
+func (v *validator) sync() {
+	// get a copy of the CRL map, the contents should not be changed directly.
+	crlMap := v.getCRLs()
 
-	for _, certificate := range certificates {
-		subject := certificate.Subject.String()
+	// TODO: does this need a WaitGroup?
+	wg := &sync.WaitGroup{}
+	wg.Add(len(crlMap))
 
-		if _, ok := v.certificates[subject]; !ok {
-			v.certificates[subject] = certificate
+	// Update in parallel if at least one of the issuers is still valid
+	for endpoint, current := range crlMap {
+		if len(current.issuers) > 0 && !anyCertificateActive(current.issuers...) {
+			wg.Done()
+			continue
 		}
-	}
-}
-
-func (v *validator) VerifyPeerCertificateFunction(maxValidityDays int) func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
-	return func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
-		// Import unknown certificates
-		var verifiedCerts []*x509.Certificate
 
-		for _, chain := range verifiedChains {
-			for _, verifiedCert := range chain {
-				if verifiedCert.IsCA {
-					verifiedCerts = append(verifiedCerts, verifiedCert)
+		go func(ep string, rl *revocationList, wg *sync.WaitGroup) {
+			err := v.updateCRL(ep, rl)
+			if err != nil {
+				// Connections containing a certificate pointing to this CRL will be accepted until its current.list.NextUpdate.
+				if rl != nil && time.Since(rl.lastUpdated) > 24*time.Hour {
+					// Escalate to error logging if the CRL has not been updated in a day.
+					log.Logger().WithError(err).WithField("CRLDistributionPoint", ep).Error("Update CRL")
+				} else {
+					log.Logger().WithError(err).WithField("CRLDistributionPoint", ep).Debug("Update CRL")
 				}
 			}
-		}
+			wg.Done()
+		}(endpoint, current, wg)
+	}
 
-		v.appendCertificates(verifiedCerts)
+	wg.Wait()
+}
 
-		// Parse certificates and check if they are revoked
-		var raw []byte
+// updateCRL downloads the CRL from endpoint, and updates the revocationList in crls if it is newer than current
+func (v *validator) updateCRL(endpoint string, current *revocationList) error {
+	// download CRL
+	crl, err := v.downloadCRL(endpoint)
+	if err != nil {
+		return err
+	}
 
-		for _, rawCert := range rawCerts {
-			raw = append(raw, rawCert...)
-		}
+	// verify signature
+	if err = v.verifyCRL(crl); err != nil {
+		return err
+	}
 
-		certificates, err := x509.ParseCertificates(raw)
-		if err != nil {
-			return err
+	// update when it is a new clr
+	if current == nil || current.list.Number == nil || current.list.Number.Cmp(crl.Number) < 0 {
+		// parse revocations
+		revoked := make(map[string]*pkix.RevokedCertificate, len(crl.RevokedCertificates))
```

Comment:
No need to store `*pkix.RevokedCertificate`, cann make `revoked` just `map[string]bool`